<html>
<head>
<title>MayaChemTools:Documentation:SequenceFileUtil.pm</title>
<meta http-equiv="content-type" content="text/html;charset=utf-8">
<link rel="stylesheet" type="text/css" href="../../css/MayaChemTools.css">
</head>
<body leftmargin="20" rightmargin="20" topmargin="10" bottommargin="10">
<br/>
<center>
<a href="http://www.mayachemtools.org" title="MayaChemTools Home"><img src="../../images/MayaChemToolsLogo.gif" border="0" alt="MayaChemTools"></a>
</center>
<br/>
<div class="DocNav">
<table width="100%" border=0 cellpadding=0 cellspacing=2>
<tr align="left" valign="top"><td width="33%" align="left"><a href="./SDFileUtil.html" title="SDFileUtil.html">Previous</a>&nbsp;&nbsp;<a href="./index.html" title="Table of Contents">TOC</a>&nbsp;&nbsp;<a href="./StatisticsUtil.html" title="StatisticsUtil.html">Next</a></td><td width="34%" align="middle"><strong>SequenceFileUtil.pm</strong></td><td width="33%" align="right"><a href="././code/SequenceFileUtil.html" title="View source code">Code</a>&nbsp;|&nbsp;<a href="./../pdf/SequenceFileUtil.pdf" title="PDF US Letter Size">PDF</a>&nbsp;|&nbsp;<a href="./../pdfgreen/SequenceFileUtil.pdf" title="PDF US Letter Size with narrow margins: www.changethemargins.com">PDFGreen</a>&nbsp;|&nbsp;<a href="./../pdfa4/SequenceFileUtil.pdf" title="PDF A4 Size">PDFA4</a>&nbsp;|&nbsp;<a href="./../pdfa4green/SequenceFileUtil.pdf" title="PDF A4 Size with narrow margins: www.changethemargins.com">PDFA4Green</a></td></tr>
</table>
</div>
<h1 id="NAME">NAME</h2>
<p>SequenceFileUtil</p>
<h1 id="SYNOPSIS">SYNOPSIS</h2>
<p>use SequenceFileUtil ;</p>
<p>use SequenceFileUtil qw(:all);</p>
<h1 id="DESCRIPTION">DESCRIPTION</h2>
<p><b>SequenceFileUtil</b> module provides the following functions:</p>
<p>
 <a href="#AreSequenceLengthsIdentical">AreSequenceLengthsIdentical</a>, <a href="#CalcuatePercentSequenceIdentity">CalcuatePercentSequenceIdentity</a>, <a href="#CalculatePercentSequenceIdentityMatrix">CalculatePercentSequenceIdentityMatrix</a>, <a href="#GetLongestSequence">GetLongestSequence</a>, <a href="#GetSequenceLength">GetSequenceLength</a>, <a href="#GetShortestSequence">GetShortestSequence</a>, <a href="#IsClustalWSequenceFile">IsClustalWSequenceFile</a>, <a href="#IsGapResidue">IsGapResidue</a>, <a href="#IsMSFSequenceFile">IsMSFSequenceFile</a>, <a href="#IsPIRFastaSequenceFile">IsPIRFastaSequenceFile</a>, <a href="#IsPearsonFastaSequenceFile">IsPearsonFastaSequenceFile</a>, <a href="#IsSupportedSequenceFile">IsSupportedSequenceFile</a>, <a href="#ReadClustalWSequenceFile">ReadClustalWSequenceFile</a>, <a href="#ReadMSFSequenceFile">ReadMSFSequenceFile</a>, <a href="#ReadPIRFastaSequenceFile">ReadPIRFastaSequenceFile</a>, <a href="#ReadPearsonFastaSequenceFile">ReadPearsonFastaSequenceFile</a>, <a href="#ReadSequenceFile">ReadSequenceFile</a>, <a href="#RemoveSequenceAlignmentGapColumns">RemoveSequenceAlignmentGapColumns</a>, <a href="#RemoveSequenceGaps">RemoveSequenceGaps</a>, <a href="#WritePearsonFastaSequenceFile SequenceFileUtil module provides various methods to process sequence files and retreive appropriate information.">WritePearsonFastaSequenceFile SequenceFileUtil module provides various methods to process sequence files and retreive appropriate information.</a>
</p><h1 id="FUNCTIONS">FUNCTIONS</h2>
<dl>
<dt id="AreSequenceLengthsIdentical"><b>AreSequenceLengthsIdentical</b></dt>
<dd>
<div class="OptionsBox">    $Status = AreSequenceLengthsIdentical($SequencesDataRef);</div>
<p>Checks the lengths of all the sequences available in <i>SequencesDataRef</i> and returns 1 or 0 based whether lengths of all the sequence is same.</p>
</dd>
<dt id="CalcuatePercentSequenceIdentity"><b>CalcuatePercentSequenceIdentity</b></dt>
<dd>
<div class="OptionsBox">    $PercentIdentity =
       AreSequenceLengthsIdenticalAreSequenceLengthsIdentical(
          $Sequence1, $Sequence2, [$IgnoreGaps, $Precision]);</div>
<p>Returns percent identity between <i>Sequence1</i> and <i>Sequence2</i>. Optional arguments <i>IgnoreGaps</i> and <i>Precision</i> control handling of gaps in sequences and precision of the returned value. By default, gaps are ignored and precision is set up to 1 decimal.</p>
</dd>
<dt id="CalculatePercentSequenceIdentityMatrix"><b>CalculatePercentSequenceIdentityMatrix</b></dt>
<dd>
<div class="OptionsBox">    $IdentityMatrixDataRef = CalculatePercentSequenceIdentityMatrix(
                             $SequencesDataRef, [$IgnoreGaps,
                             $Precision]);</div>
<p>Calculate pairwise percent identity between all the sequences available in <i>SequencesDataRef</i> and returns a reference to identity matrix hash. Optional arguments <i>IgnoreGaps</i> and <i>Precision</i> control handling of gaps in sequences and precision of the returned value. By default, gaps are ignored and precision is set up to 1 decimal.</p>
</dd>
<dt id="GetSequenceLength"><b>GetSequenceLength</b></dt>
<dd>
<div class="OptionsBox">    $SeqquenceLength = GetSequenceLength($Sequence, [$IgnoreGaps]);</div>
<p>Returns length of the specified sequence. Optional argument <i>IgnoreGaps</i> controls handling of gaps. By default, gaps are ignored.</p>
</dd>
<dt id="GetShortestSequence"><b>GetShortestSequence</b></dt>
<dd>
<div class="OptionsBox">   ($ID, $Sequence, $SeqLen, $Description) = GetShortestSequence(
&nbsp;&nbsp;&nbsp;       $SequencesDataRef, [$IgnoreGaps]);</div>
<p>Checks the lengths of all the sequences available in $SequencesDataRef and returns $ID, $Sequence, $SeqLen, and $Description values for the shortest sequence. Optional arguments $IgnoreGaps controls handling of gaps in sequences. By default, gaps are ignored.</p>
</dd>
<dt id="GetLongestSequence"><b>GetLongestSequence</b></dt>
<dd>
<div class="OptionsBox">   ($ID, $Sequence, $SeqLen, $Description) = GetLongestSequence(
&nbsp;&nbsp;&nbsp;       $SequencesDataRef, [$IgnoreGaps]);</div>
<p>Checks the lengths of all the sequences available in <i>SequencesDataRef</i> and returns <b>ID</b>, <b>Sequence</b>, <b>SeqLen</b>, and <b>Description</b> values for the longest sequence. Optional argument $<i>IgnoreGaps</i> controls handling of gaps in sequences. By default, gaps are ignored.</p>
</dd>
<dt id="IsGapResidue"><b>IsGapResidue</b></dt>
<dd>
<div class="OptionsBox">    $Status = AreSequenceLengthsIdentical($Residue);</div>
<p>Returns 1 or 0 based on whether <i>Residue</i> corresponds to a gap. Any character other than A to Z is considered a gap residue.</p>
</dd>
<dt id="IsSupportedSequenceFile"><b>IsSupportedSequenceFile</b></dt>
<dd>
<div class="OptionsBox">    $Status = IsSupportedSequenceFile($SequenceFile);</div>
<p>Returns 1 or 0 based on whether <i>SequenceFile</i> corresponds to a supported sequence format.</p>
</dd>
<dt id="IsClustalWSequenceFile"><b>IsClustalWSequenceFile</b></dt>
<dd>
<div class="OptionsBox">    $Status = IsClustalWSequenceFile($SequenceFile);</div>
<p>Returns 1 or 0 based on whether <i>SequenceFile</i> corresponds to Clustal sequence alignment format.</p>
</dd>
<dt id="IsPearsonFastaSequenceFile"><b>IsPearsonFastaSequenceFile</b></dt>
<dd>
<div class="OptionsBox">    $Status = IsPearsonFastaSequenceFile($SequenceFile);</div>
<p>Returns 1 or 0 based on whether <i>SequenceFile</i> corresponds to Pearson FASTA sequence format.</p>
</dd>
<dt id="IsPIRFastaSequenceFile"><b>IsPIRFastaSequenceFile</b></dt>
<dd>
<div class="OptionsBox">    $Status = IsPIRFastaSequenceFile($SequenceFile);</div>
<p>Returns 1 or 0 based on whether <i>SequenceFile</i> corresponds to PIR FASTA sequence format.</p>
</dd>
<dt id="IsMSFSequenceFile"><b>IsMSFSequenceFile</b></dt>
<dd>
<div class="OptionsBox">    $Status = IsClustalWSequenceFile($SequenceFile);</div>
<p>Returns 1 or 0 based on whether <i>SequenceFile</i> corresponds to MSF sequence alignment format.</p>
</dd>
<dt id="ReadSequenceFile"><b>ReadSequenceFile</b></dt>
<dd>
<div class="OptionsBox">    $SequenceDataMapRef = ReadSequenceFile($SequenceFile);</div>
<p>Reads <i>SequenceFile</i> and returns reference to a hash containing following key/value pairs:</p>
<div class="OptionsBox">    $SequenceDataMapRef-&gt;{IDs} - Array of sequence IDs
<br/>    $SequenceDataMapRef-&gt;{Count} - Number of sequences
<br/>    $SequenceDataMapRef-&gt;{Description}{$ID} - Sequence description
<br/>    $SequenceDataMapRef-&gt;{Sequence}{$ID} - Sequence for a specific ID
<br/>    $SequenceDataMapRef-&gt;{Sequence}{InputFileType} - File format</div>
</dd>
<dt id="ReadClustalWSequenceFile"><b>ReadClustalWSequenceFile</b></dt>
<dd>
<div class="OptionsBox">    $SequenceDataMapRef = ReadClustalWSequenceFile($SequenceFile);</div>
<p>Reads ClustalW <i>SequenceFile</i> and returns reference to a hash containing following key/value pairs as describes in <b>ReadSequenceFile</b> method.</p>
</dd>
<dt id="ReadMSFSequenceFile"><b>ReadMSFSequenceFile</b></dt>
<dd>
<div class="OptionsBox">    $SequenceDataMapRef = ReadMSFSequenceFile($SequenceFile);</div>
<p>Reads MSF <i>SequenceFile</i> and returns reference to a hash containing following key/value pairs as describes in <b>ReadSequenceFile</b> method.</p>
</dd>
<dt id="ReadPIRFastaSequenceFile"><b>ReadPIRFastaSequenceFile</b></dt>
<dd>
<div class="OptionsBox">    $SequenceDataMapRef = ReadPIRFastaSequenceFile($SequenceFile);</div>
<p>Reads PIR FASTA <i>SequenceFile</i> and returns reference to a hash containing following key/value pairs as describes in <b>ReadSequenceFile</b> method.</p>
</dd>
<dt id="ReadPearsonFastaSequenceFile"><b>ReadPearsonFastaSequenceFile</b></dt>
<dd>
<div class="OptionsBox">    $SequenceDataMapRef = ReadPearsonFastaSequenceFile($SequenceFile);</div>
<p>Reads Pearson FASTA <i>SequenceFile</i> and returns reference to a hash containing following key/value pairs as describes in <b>ReadSequenceFile</b> method.</p>
</dd>
<dt id="RemoveSequenceGaps"><b>RemoveSequenceGaps</b></dt>
<dd>
<div class="OptionsBox">    $SeqWithoutGaps = RemoveSequenceGaps($Sequence);</div>
<p>Removes gaps from <i>Sequence</i> and return a sequence without any gaps.</p>
</dd>
<dt id="RemoveSequenceAlignmentGapColumns"><b>RemoveSequenceAlignmentGapColumns</b></dt>
<dd>
<div class="OptionsBox">    $NewAlignmentDataMapRef = RemoveSequenceAlignmentGapColumns(
                              $AlignmentDataMapRef);</div>
<p>Using input alignment data map ref containing following keys, generate a new hash with same set of keys after residue columns containg only gaps have been removed:</p>
<div class="OptionsBox">    {IDs} : Array of IDs in order as they appear in file
<br/>    {Count}: ID count
<br/>    {Description}{$ID} : Description data
<br/>    {Sequence}{$ID} : Sequence data</div>
</dd>
<dt id="WritePearsonFastaSequenceFile"><b>WritePearsonFastaSequenceFile</b></dt>
<dd>
<div class="OptionsBox">    WritePearsonFastaSequenceFile($SequenceFileName, $SequenceDataRef,
                                  [$MaxLength]);</div>
<p>Using sequence data specified via <i>SequenceDataRef</i>, write out a Pearson FASTA sequence file. Optional argument <i>MaxLength</i> controls maximum length sequence in each line; default is 80.</p>
</dd>
</dl>
<h1 id="AUTHOR">AUTHOR</h2>
<p><a href="mailto:msud@san.rr.com">Manish Sud</a></p>
<h1 id="SEE-ALSO">SEE ALSO</h2>
<p><a href="./PDBFileUtil.html">PDBFileUtil.pm</a></p>
<h1 id="COPYRIGHT">COPYRIGHT</h2>
<p>Copyright (C) 2018 Manish Sud. All rights reserved.</p>
<p>This file is part of MayaChemTools.</p>
<p>MayaChemTools is free software; you can redistribute it and/or modify it under the terms of the GNU Lesser General Public License as published by the Free Software Foundation; either version 3 of the License, or (at your option) any later version.</p>
<p>&nbsp</p><p>&nbsp</p><div class="DocNav">
<table width="100%" border=0 cellpadding=0 cellspacing=2>
<tr align="left" valign="top"><td width="33%" align="left"><a href="./SDFileUtil.html" title="SDFileUtil.html">Previous</a>&nbsp;&nbsp;<a href="./index.html" title="Table of Contents">TOC</a>&nbsp;&nbsp;<a href="./StatisticsUtil.html" title="StatisticsUtil.html">Next</a></td><td width="34%" align="middle"><strong>May 15, 2018</strong></td><td width="33%" align="right"><strong>SequenceFileUtil.pm</strong></td></tr>
</table>
</div>
<br />
<center>
<img src="../../images/h2o2.png">
</center>
</body>
</html>
