<html>
<head>
<title>MayaChemTools:Documentation:AnalyzeSequenceFilesData.pl</title>
<meta http-equiv="content-type" content="text/html;charset=utf-8">
<link rel="stylesheet" type="text/css" href="../../css/MayaChemTools.css">
</head>
<body leftmargin="20" rightmargin="20" topmargin="10" bottommargin="10">
<br/>
<center>
<a href="http://www.mayachemtools.org" title="MayaChemTools Home"><img src="../../images/MayaChemToolsLogo.gif" border="0" alt="MayaChemTools"></a>
</center>
<br/>
<div class="DocNav">
<table width="100%" border=0 cellpadding=0 cellspacing=2>
<tr align="left" valign="top"><td width="33%" align="left"><a href="./AnalyzeSDFilesData.html" title="AnalyzeSDFilesData.html">Previous</a>&nbsp;&nbsp;<a href="./index.html" title="Table of Contents">TOC</a>&nbsp;&nbsp;<a href="./AnalyzeTextFilesData.html" title="AnalyzeTextFilesData.html">Next</a></td><td width="34%" align="middle"><strong>AnalyzeSequenceFilesData.pl</strong></td><td width="33%" align="right"><a href="././code/AnalyzeSequenceFilesData.html" title="View source code">Code</a>&nbsp;|&nbsp;<a href="./../pdf/AnalyzeSequenceFilesData.pdf" title="PDF US Letter Size">PDF</a>&nbsp;|&nbsp;<a href="./../pdfgreen/AnalyzeSequenceFilesData.pdf" title="PDF US Letter Size with narrow margins: www.changethemargins.com">PDFGreen</a>&nbsp;|&nbsp;<a href="./../pdfa4/AnalyzeSequenceFilesData.pdf" title="PDF A4 Size">PDFA4</a>&nbsp;|&nbsp;<a href="./../pdfa4green/AnalyzeSequenceFilesData.pdf" title="PDF A4 Size with narrow margins: www.changethemargins.com">PDFA4Green</a></td></tr>
</table>
</div>
<h1 id="NAME">NAME</h2>
<p>AnalyzeSequenceFilesData.pl - Analyze sequence and alignment files</p>
<h1 id="SYNOPSIS">SYNOPSIS</h2>
<p>AnalyzeSequenceFilesData.pl SequenceFile(s) AlignmentFile(s)...</p>
<p>AnalyzeSequenceFilesData.pl [<b>-h, --help</b>] [<b>-i, --IgnoreGaps</b> yes | no] [<b>-m, --mode</b> PercentIdentityMatrix | ResidueFrequencyAnalysis | All] [<b>--outdelim</b> comma | tab | semicolon] [<b>-o, --overwrite</b>] [<b>-p, --precision</b> number] [<b>-q, --quote</b> yes | no] [<b>--ReferenceSequence</b> SequenceID | UseFirstSequenceID] [<b>--region</b> &quot;StartResNum, EndResNum, [StartResNum, EndResNum...]&quot; | UseCompleteSequence] [<b>--RegionResiduesMode</b> AminoAcids | NucleicAcids | None] [<b>-w, --WorkingDir</b> dirname] SequenceFile(s) AlignmentFile(s)...</p>
<h1 id="DESCRIPTION">DESCRIPTION</h2>
<p>Analyze <i>SequenceFile(s) and AlignmentFile(s)</i> data: calculate pairwise percent identity matrix or calculate percent occurrence of various residues in specified sequence regions. All the sequences in the input file must have the same sequence lengths; otherwise, the sequence file is ignored.</p>
<p>The file names are separated by spaces. All the sequence files in a current directory can be specified by <i>*.aln</i>, <i>*.msf</i>, <i>*.fasta</i>, <i>*.fta</i>, <i>*.pir</i> or any other supported formats; additionally, <i>DirName</i> corresponds to all the sequence files in the current directory with any of the supported file extension: <i>.aln, .msf, .fasta, .fta, and .pir</i>.</p>
<p>Supported sequence formats are: <i>ALN/CLustalW</i>, <i>GCG/MSF</i>, <i>PILEUP/MSF</i>, <i>Pearson/FASTA</i>, and <i>NBRF/PIR</i>. Instead of using file extensions, file formats are detected by parsing the contents of <i>SequenceFile(s) and AlignmentFile(s)</i>.</p>
<h1 id="OPTIONS">OPTIONS</h2>
<dl>
<dt id="h---help"><b>-h, --help</b></dt>
<dd>
<p>Print this help message.</p>
</dd>
<dt id="i---IgnoreGaps-yes-no"><b>-i, --IgnoreGaps</b> <i>yes | no</i></dt>
<dd>
<p>Ignore gaps during calculation of sequence lengths and specification of regions during residue frequency analysis. Possible values: <i>yes or no</i>. Default value: <i>yes</i>.</p>
</dd>
<dt id="m---mode-PercentIdentityMatrix-ResidueFrequencyAnalysis-All"><b>-m, --mode</b> <i>PercentIdentityMatrix | ResidueFrequencyAnalysis | All</i></dt>
<dd>
<p>Specify how to analyze data in sequence files: calculate percent identity matrix or calculate frequency of occurrence of residues in specific regions. During <i>ResidueFrequencyAnalysis</i> value of <b>-m, --mode</b> option, output files are generated for both the residue count and percent residue count. Possible values: <i>PercentIdentityMatrix, ResidueFrequencyAnalysis, or All</i>. Default value: <i>PercentIdentityMatrix</i>.</p>
</dd>
<dt id="outdelim-comma-tab-semicolon"><b>--outdelim</b> <i>comma | tab | semicolon</i></dt>
<dd>
<p>Output text file delimiter. Possible values: <i>comma, tab, or semicolon</i>. Default value: <i>comma</i>.</p>
</dd>
<dt id="o---overwrite"><b>-o, --overwrite</b></dt>
<dd>
<p>Overwrite existing files.</p>
</dd>
<dt id="p---precision-number"><b>-p, --precision</b> <i>number</i></dt>
<dd>
<p>Precision of calculated values in the output file. Default: up to <i>2</i> decimal places. Valid values: positive integers.</p>
</dd>
<dt id="q---quote-yes-no"><b>-q, --quote</b> <i>yes | no</i></dt>
<dd>
<p>Put quotes around column values in output text file. Possible values: <i>yes or no</i>. Default value: <i>yes</i>.</p>
</dd>
<dt id="ReferenceSequence-SequenceID-UseFirstSequenceID"><b>--ReferenceSequence</b> <i>SequenceID | UseFirstSequenceID</i></dt>
<dd>
<p>Specify reference sequence ID to identify regions for performing <i>ResidueFrequencyAnalysis</i> specified using <b>-m, --mode</b> option. Default: <i>UseFirstSequenceID</i>.</p>
</dd>
<dt id="region-StartResNum-EndResNum-StartResNum-EndResNum...-UseCompleteSequence"><b>--region</b> <i>StartResNum,EndResNum,[StartResNum,EndResNum...] | UseCompleteSequence</i></dt>
<dd>
<p>Specify how to perform frequency of occurrence analysis for residues: use specific regions indicated by starting and ending residue numbers in reference sequence or use the whole reference sequence as one region. Default: <i>UseCompleteSequence</i>.</p>
<p>Based on the value of <b>-i, --IgnoreGaps</b> option, specified residue numbers <i>StartResNum,EndResNum</i> correspond to the positions in the reference sequence without gaps or with gaps.</p>
<p>For residue numbers corresponding to the reference sequence including gaps, percent occurrence of various residues corresponding to gap position in reference sequence is also calculated.</p>
</dd>
<dt id="RegionResiduesMode-AminoAcids-NucleicAcids-None"><b>--RegionResiduesMode</b> <i>AminoAcids | NucleicAcids | None</i></dt>
<dd>
<p>Specify how to process residues in the regions specified using <b>--region</b> option during <i>ResidueFrequencyAnalysis</i> calculation: categorize residues as amino acids, nucleic acids, or simply ignore residue category during the calculation. Possible values: <i>AminoAcids, NucleicAcids or None</i>. Default value: <i>None</i>.</p>
<p>For <i>AminoAcids</i> or <i>NucleicAcids</i> values of <b>--RegionResiduesMode</b> option, all the standard amino acids or nucleic acids are listed in the output file for each region; Any gaps and other non standard residues are added to the list as encountered.</p>
<p>For <i>None</i> value of <b>--RegionResiduesMode</b> option, no assumption is made about type of residues. Residue and gaps are added to the list as encountered.</p>
</dd>
<dt id="r---root-rootname"><b>-r, --root</b> <i>rootname</i></dt>
<dd>
<p>New sequence file name is generated using the root: &lt;Root&gt;&lt;Mode&gt;.&lt;Ext&gt; and &lt;Root&gt;&lt;Mode&gt;&lt;RegionNum&gt;.&lt;Ext&gt;. Default new file name: &lt;SequenceFileName&gt;&lt;Mode&gt;.&lt;Ext&gt; for <i>PercentIdentityMatrix</i> value <b>m, --mode</b> option and &lt;SequenceFileName&gt;&lt;Mode&gt;&lt;RegionNum&gt;.&lt;Ext&gt; for <i>ResidueFrequencyAnalysis</i>. The csv, and tsv &lt;Ext&gt; values are used for comma/semicolon, and tab delimited text files respectively. This option is ignored for multiple input files.</p>
</dd>
<dt id="w---WorkingDir-text"><b>-w --WorkingDir</b> <i>text</i></dt>
<dd>
<p>Location of working directory. Default: current directory.</p>
</dd>
</dl>
<h1 id="EXAMPLES">EXAMPLES</h2>
<p>To calculate percent identity matrix for all sequences in Sample1.msf file and generate Sample1PercentIdentityMatrix.csv, type:</p>
<div class="ExampleBox">    % AnalyzeSequenceFilesData.pl Sample1.msf</div>
<p>To perform residue frequency analysis for all sequences in Sample1.aln file corresponding to non-gap positions in the first sequence and generate Sample1ResidueFrequencyAnalysisRegion1.csv and Sample1PercentResidueFrequencyAnalysisRegion1.csv files, type:</p>
<div class="ExampleBox">    % AnalyzeSequenceFilesData.pl -m ResidueFrequencyAnalysis -o
      Sample1.aln</div>
<p>To perform residue frequency analysis for all sequences in Sample1.aln file corresponding to all positions in the first sequence and generate TestResidueFrequencyAnalysisRegion1.csv and TestPercentResidueFrequencyAnalysisRegion1.csv files, type:</p>
<div class="ExampleBox">    % AnalyzeSequenceFilesData.pl -m ResidueFrequencyAnalysis --IgnoreGaps
      No -o -r Test Sample1.aln</div>
<p>To perform residue frequency analysis for all sequences in Sample1.aln file corresponding to non-gap residue positions 5 to 10, and 30 to 40 in sequence ACHE_BOVIN and generate Sample1ResidueFrequencyAnalysisRegion1.csv, Sample1ResidueFrequencyAnalysisRegion2.csv, SamplePercentResidueFrequencyAnalysisRegion1.csv, and SamplePercentResidueFrequencyAnalysisRegion2.csv files, type:</p>
<div class="ExampleBox">    % AnalyzeSequenceFilesData.pl -m ResidueFrequencyAnalysis
      --ReferenceSequence ACHE_BOVIN --region &quot;5,15,30,40&quot; -o Sample1.msf</div>
<h1 id="AUTHOR">AUTHOR</h2>
<p><a href="mailto:msud@san.rr.com">Manish Sud</a></p>
<h1 id="SEE-ALSO">SEE ALSO</h2>
<p><a href="./ExtractFromSequenceFiles.html">ExtractFromSequenceFiles.pl</a>, <a href="./InfoSequenceFiles.html">InfoSequenceFiles.pl</a></p>
<h1 id="COPYRIGHT">COPYRIGHT</h2>
<p>Copyright (C) 2018 Manish Sud. All rights reserved.</p>
<p>This file is part of MayaChemTools.</p>
<p>MayaChemTools is free software; you can redistribute it and/or modify it under the terms of the GNU Lesser General Public License as published by the Free Software Foundation; either version 3 of the License, or (at your option) any later version.</p>
<p>&nbsp</p><p>&nbsp</p><div class="DocNav">
<table width="100%" border=0 cellpadding=0 cellspacing=2>
<tr align="left" valign="top"><td width="33%" align="left"><a href="./AnalyzeSDFilesData.html" title="AnalyzeSDFilesData.html">Previous</a>&nbsp;&nbsp;<a href="./index.html" title="Table of Contents">TOC</a>&nbsp;&nbsp;<a href="./AnalyzeTextFilesData.html" title="AnalyzeTextFilesData.html">Next</a></td><td width="34%" align="middle"><strong>May 15, 2018</strong></td><td width="33%" align="right"><strong>AnalyzeSequenceFilesData.pl</strong></td></tr>
</table>
</div>
<br />
<center>
<img src="../../images/h2o2.png">
</center>
</body>
</html>
